### PR TITLE
fix: universal link regex

### LINKED ISSUE #17869: Incorrect status link doesn't open in Status browser
### Steps:
1. Navigate to an incorrect Status link. (Example: https://status.app/cffc/Ow==#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11)
2. Select 'Open in Status' option

### Actual result:
The webview is not opened in the browser when the broken Status link is tapped.

https://github.com/status-im/status-mobile/assets/52490791/aa780556-8920-447b-b3cc-189d1710efd9

### Expected result:
The user should be navigated to the webview when tapping an incorrect Status link.

### ENV:
Nightly [10 Nov 2023]

### PR DESCRIPTION
**Pull request by yqrashawn**
fixes #17869

-------

[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

### Summary

update universal link regex

issue caused by bad universal link regex used when open URL in built-in browser

thus causing this loop
URL -> handle universal-link -> fallback to open in browser -> oh it's universal link -> handler universal-link

### Testing notes
affects places using this regex
- src/status_im/browser/core.cljs -> all kinds of link handler functions
- src/status_im/wallet/choose_recipient/core.cljs -> `:wallet-legacy/parse-eip681-uri-and-resolve-ens`

status: ready

### CODE REVIEW

**Review by alwx — APPROVED** [review]

**Review by BalogunofAfrica — APPROVED** [review]

**Comment by yqrashawn** [comment]
rebased onto dev

**Comment by qoqobolo** [comment]
Hi @yqrashawn , thanks for the fix!
Have a couple of questions

### Q1: Nothing happens when you click on an invalid link if you add extra characters to the key (to make it invalid)

watch the video please:


https://github.com/status-im/status-mobile/assets/67952253/9e433678-880b-4ccd-8fda-a90ecbb9e538

1. first I send a valid link `https://status.app/u#zQ3sheBj4d9C7uT4EQT76opXhkzLVRh7H2gXQvRuf4ETSvfqZ`
2. then I add extra characters to the `/u#` part and the fix is working here - I'm redirected to the browser and get a 404 page 
3. then I add extra characters at the end of the link and when I click it nothing happens.
What's the expected behavior in that case?

### Q2: could you give me a clue on testing notes? :)

> Testing notes
> affects places using this regex

> src/status_im/browser/core.cljs -> all kinds of link handler functions
> src/status_im/wallet/choose_recipient/core.cljs -> :wallet-legacy/parse-eip681-uri-and-resolve-ens

What should I pay attention to besides, for example, scanning such links in the new wallet?
Screenshot 2023-12-05 at 17 48 51

**Comment by yqrashawn** [comment]
Q1: fixed now
Q2: 
I think that's good enough.
That logic only happens when URL starts with `ethereum:`

**Comment by qoqobolo** [comment]
Thanks @yqrashawn 

### ISSUE 1: Some link formats no longer trigger app launch on Android

When you click the following links in an external messenger you get 404 on web or `Application error: a client-side exception has occured`. The same links work fine on nightly 

https://status.app/u#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj
https://status.app/c#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11
https://status.app/u/Ow==#zQ3shsKnV5HJMWJR61c6dssWzHszdbLfBoMF1gcLtSQAYdw2d
https://status.app/c/Ow==#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11


PR

https://github.com/status-im/status-mobile/assets/67952253/ca71c9f9-dea6-44dd-bab5-a5cc2d55a5d3


https://github.com/status-im/status-mobile/assets/67952253/cfef6af5-757a-48e6-a3fd-b745240e3c08

---

Nightly


https://github.com/status-im/status-mobile/assets/67952253/698531b8-5aa1-4ef1-b25b-546ce890f689

Links of this format also work fine: https://status.app/u/G10A4B0JdgwyRww90WXtnP1oNH1ZLQNM0yX0Ja9YyAMjrqSZIYINOHCbFhrnKRAcPGStPxCMJDSZlGCKzmZrJcimHY8BbcXlORrElv_BbQEegnMDPx1g9C5VVNl0fE4y#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj

And all the formats redirect you to the right place if they are posted inside the app in a chat. 

**Steps to reproduce:**
1. Install PR build
2. Follow the link from an external messenger (Telegram, etc), for example https://status.app/u#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj

**Comment by qoqobolo** [comment]
### Q3: Why if I add extra characters to the `/u#` part I land on the web page with the new design (with animation) and if I add extra characters at the end of the link I land on the old page (`Oh no something went wrong!`) ?

Is it expected?


https://github.com/status-im/status-mobile/assets/67952253/fd5fc16b-b3e3-4711-8190-2718e598a510

**Comment by qoqobolo** [comment]
@yqrashawn I forgot to mention yesterday that Issue 1 is only relevant for Android. Updated the description. Pinging you so you don’t miss it if you’ve already started working on it

**Comment by yqrashawn** [comment]
Hi @qoqobolo 

https://github.com/status-im/status-mobile/assets/15090582/2158bfa8-dffd-4507-9152-f693401702ea

For ISSUE 1
I tested debug and PR build on real device and the links work fine when opening in Chrome. 
What's the app you were using opening universal links in the videos?

What's more, the changes in this PR won't affect the universal link trigger app open 
since the logic happens before the app even got opened
it's the configurations in the `android` folder that affects this behavior

For Q3
> Q3: Why if I add extra characters to the /u# part I land on the web page with the new design (with animation) and if I add extra characters at the end of the link I land on the old page (Oh no something went wrong!) ?

I guess that's similar route issue happens to status-website, some links matched to the backend handler and some matched to the client handler. @felicio knows better

**Comment by qoqobolo** [comment]
> What's the app you were using opening universal links in the videos?

I'm using Telegram where I open your page with links https://yqrashawn.com/_test/status-universal-link.html


https://github.com/status-im/status-mobile/assets/67952253/446c12d1-9c9c-4f75-a2b0-0eff63eada79

**Comment by felicio** [comment]
@yqrashawn @qoqobolo 

> Q3: Why if I add extra characters to the /u# part I land on the web page with the new design (with animation) and if I add extra characters at the end of the link I land on the old page (Oh no something went wrong!) ?

I'd like to help best as I can, however some videos without the repro links just like the ones that are being clicked or as they appear in the browser's address bar can be hard to follow.

BUT,

> Why if I add extra characters to the /u# part I land on the web page with the new design (with animation)

For example, /uBlah is an unsupported route, hence the 404/not found page. And that's expected.

>  and if I add extra characters at the end of the link I land on the old page (Oh no something went wrong!) ?

Technically, it's not an old page and it's not 404, but 500/error/unsuccessful data decoding  instead. And if you modify the chat key for example, it would yield this error. To be fair, the error could be more descriptive (i.e. Invalid public key) and later I'll attach a non-blocking issue addressing that to this message.

As far as the redirecting goes, I understand and I am okay with for the status apps to try to handle only the routes it does some work on (i.e. displays dialog or navigates to /u/, /c/, /cc/  internally) and redirect rest.

HOWEVER,

@yqrashawn  I'd like to know your opinion on, if the apps in case of supported routes like mentioned above shouldn't display the errors (e.g. invalid public key)  internally too (e.g. in place of the profile/community/channel dialog), instead of redirecting to website.

---

More descriptive website error: https://github.com/status-im/status-website/issues/702 (non-blocking)

**Comment by felicio** [comment]
@qoqobolo 

> https://status.app/c/Ow==#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11

The "Application error" for this link is a bit something different and I'll have a look.

---

Handle missing required fields: https://github.com/status-im/status-website/issues/703 (non-blocking)

**Comment by yqrashawn** [comment]
> As far as the redirecting goes, I understand and I am okay with ...
I'm fine with it as well.

**Comment by felicio** [comment]
@yqrashawn 

> > As far as the redirecting goes, I understand and I am okay with ...
> > I'm fine with it as well.

Cool 👍, that's the case of unknown routes.

>  if the apps in case of supported routes like mentioned above shouldn't display the errors (e.g. invalid public key) internally too (e.g. in place of the profile/community/channel dialog), instead of redirecting to website.

Please, did you have time to consider the suggestion without redirecting for known routes but with some invalid data (pk here)  as well?

**Comment by qoqobolo** [comment]
> I'd like to help best as I can, however some videos without the repro links just like the ones that are being clicked or as they appear in the browser's address bar can be hard to follow.

Oh sorry, I should've posted some examples of links for sure, completely agree that it would be more convenient in that question as well.

@felicio thanks for your explanation and the logged issues, now the behavior is crystal clear to me, and I agree that a redirect in the case of unknown routes is enough for now.

As for displaying internal errors in case of _supported routes but invalid encoded data_, do I understand correctly that even if we decide that we should still display errors internally, this is out of scope of this PR and will have to be logged and fixed separately?

In this case, I will wait for the results of e2e rerun, review them, and the PR will be ready to be merged. So far everything looks good to me.

**Comment by felicio** [comment]
>  even if we decide that we should still display errors internally, this is out of scope of this PR and will have to be logged and fixed separately?

You guys can decide that better than I can 🙂. I'm not sure what the latest state of "opening and displaying info of users you don't have in contacts or of communities you're not a member of internally", for example, is.

For that, I'm following https://github.com/status-im/status-desktop/issues/12245. Maybe the dialogs for that are ready and it'd be an easy change as it's related.

But yes, I'd say a follow-up is fine too, if you agree with the suggestion in the first place.

**Comment by qoqobolo** [comment]
> You guys can decide that better than I can 🙂

Yeah was about to add that that's more the question to @yqrashawn :)
I'm also not sure if we have designs for this, will check. Thanks for the answer

**Comment by qoqobolo** [comment]
@yqrashawn Found another issue with another type of invalid links, with which [the original bug](https://github.com/status-im/status-mobile/issues/17869) is still reproducible. Please take a look at this and also [at this comment](https://github.com/status-im/status-mobile/pull/18020#issuecomment-1851589010) and the following ones in case you missed it

### ISSUE 2: You get the loop when posting two links without a space between them so they form one invalid link

**Steps:**
1. Post this link in a Status chat: https://status.app/c/Ow==#zQ3shbmfT3hvh4mKa1v6uAjjyztQEroh8Mfn6Ckegjd7LT3XKhttps://status.app/c#zQ3shbmfT3hvh4mKa1v6uAjjyztQEroh8Mfn6Ckegjd7LT3XK
2. Press this link

Can we cover this case as well?

https://github.com/status-im/status-mobile/assets/67952253/6bb8ff16-43b4-4c48-bd80-123c11c9eda3

**Comment by yqrashawn** [comment]
ISSUE2 fixed

can reproduce ISSUE1 on a real device, 
universal link in telegram inner browser can't trigger app open in PR build, can trigger app open in nightly build

**Comment by yqrashawn** [comment]
Hi @qoqobolo 

I just tried latest PR build from another PR https://github.com/status-im/status-mobile/pull/18104#issuecomment-1843545685 `#15` and ISSUE 1 happens in that build as well
I guess there are some build or publish configs causing this. 
Create another issue for this https://github.com/status-im/status-mobile/issues/18177

**Comment by qoqobolo** [comment]
> ISSUE2 fixed

Thanks for the fix! But now we have another issue

### ISSUE 3: A valid link posted in a Status chat is opened in the browser

**Steps:**
1. Post a valid link in a chat (e.g. https://status.app/u/G10A4B0JdgwyRww90WXtnP1oNH1ZLQNM0yX0Ja9YyAMjrqSZIYINOHCbFhrnKRAcPGStPxCMJDSZlGCKzmZrJcimHY8BbcXlORrElv_BbQEegnMDPx1g9C5VVNl0fE4y#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj )
2. Press the link

**Expected result:** profile is opened
**Actual result:** you're asked if you want to open the link in your status or Android browser > you are redirected to the web


On the video I first open the link in nightly (expected result) and then in PR (actual result)

https://github.com/status-im/status-mobile/assets/67952253/d2e05070-bacc-47a9-9858-7b7df06b9568

**Comment by yqrashawn** [comment]
ISSUE 3 should be fixed now @qoqobolo

**Comment by qoqobolo** [comment]
@yqrashawn thank you for the fix!
Unfortunately, the issue is still reproducible, but now only for communities created on mobile (PR builds)

### ISSUE 4: Community link (`Request to join required` type) is opened in the browser if the community is created on mobile

A real user won't have the option to create a community on mobile, but we still need it for e2e tests. So please take a look at this issue

**Steps:**
1. Create a `Closed community` (which is equal to `Request to join required` type on desktop)
2. Share the community with your contact to get it posted it in a chat
3. Click the link

Link example https://status.app/c/G00AAGS9TbI9mSR-ZNmFrhRjNuEeXAAbcAIUaLLJyjMOG3ACJQ12oIHD78QhzO9s_T5bUeU7rnATWJg3mGgTUemrAg==#zQ3shYf5SquxkiY3FmCW6Nz2wuFWFcM6JEdUD62ApjAvE5YPv

**Expected result:** community screen is opened
**Actual result:** you're asked if you want to open the link in a browser 

https://github.com/status-im/status-mobile/assets/67952253/436877b6-2d03-4219-9a06-48f160810d21

**Comment by yqrashawn** [comment]
@qoqobolo ISSUE 4 should be fixed now

**Comment by qoqobolo** [comment]
@yqrashawn thanks for your work and all the fixes!
PR is ready to be merged. Please include @yevh-berdnyk 's [commit](https://github.com/status-im/status-mobile/pull/18020/commits/0fdc70de73480c5e4f97452a1d1cb01c845afd3c) when merging, it fixes e2e
